### PR TITLE
Geomech

### PR DESCRIPTION
**Pull request by hnil**
- Introduced input need for mechanical solves

### CODE REVIEW

File: src/opm/input/eclipse/EclipseState/SimulationConfig/BCConfig.cpp [hunk 1]
```diff
@@ -23,7 +23,7 @@
 #include <opm/input/eclipse/Parser/ParserKeywords/B.hpp>
 #include <opm/input/eclipse/EclipseState/SimulationConfig/BCConfig.hpp>
 #include <opm/common/utility/OpmInputError.hpp>
-
+#include <opm/input/eclipse/Schedule/BCProp.hpp>
```
Comment (atgeirr) [inline src/opm/input/eclipse/EclipseState/SimulationConfig/BCConfig.cpp:26]: This change seems unnecessary, I do not see any usage of BCProp in the file.

File: src/opm/input/eclipse/EclipseState/InitConfig/InitConfig.cpp [hunk 2]
```diff
@@ -120,6 +127,17 @@ namespace Opm {
 
         return this->equil;
     }
+    
+    bool InitConfig::hasStressEquil() const {
+        return !this->equil.empty();
```
Comment (atgeirr) [inline src/opm/input/eclipse/EclipseState/InitConfig/InitConfig.cpp:132]: Bug: should check stressequil, not equil.

File: src/opm/input/eclipse/EclipseState/InitConfig/InitConfig.cpp [hunk 3]
```diff
@@ -120,6 +127,17 @@ namespace Opm {
...
+    }
+    
+    const DeckKeyword& InitConfig::getStressEquil() const {
+        if( !this->hasStressEquil() )
+            throw std::runtime_error( "Error: No 'EQUIL' present" );
```
Comment (atgeirr) [inline src/opm/input/eclipse/EclipseState/InitConfig/InitConfig.cpp:137]: Error message is wrong (actually, due to the bug above it is correct, unintentionally...)

File: src/opm/input/eclipse/Schedule/BCProp.cpp [hunk 4]
```diff
@@ -87,6 +103,38 @@ BCProp::BCFace::BCFace(const DeckRecord& record) :
...
+    if (const auto& P = record.getItem<BCKEY::STRESSXX>(); ! P.defaultApplied(0)) {
+        mechbcvaluetmp.stress[0] = P.getSIDouble(0);
+    }
+    if (const auto& P = record.getItem<BCKEY::STRESSYY>(); ! P.defaultApplied(0)) {
+        mechbcvaluetmp.stress[1] = P.getSIDouble(0);
+    }
+    if (const auto& P = record.getItem<BCKEY::STRESSZZ>(); ! P.defaultApplied(0)) {
+        mechbcvaluetmp.stress[2] = P.getSIDouble(0);
+    }
+    mechbcvaluetmp.stress[3] = 0;
```
Comment (atgeirr) [inline src/opm/input/eclipse/Schedule/BCProp.cpp:116]: Why do we keep the three last numbers of `stress` if they are always set to zero?

**Review by atgeirr — CHANGES_REQUESTED** [review]
A few minor bugs need fixing.

If it is acceptable to those who know better than me to start using keywords with more than 8 characters, then my comments about that should be ignored. (And I would support it.)

File: src/opm/input/eclipse/share/keywords/900_OPM/B/BIOTCOEFF [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "BIOTCOEFF",
```
[the remark this answers is not in the document]
Comment (hnil) [reply src/opm/input/eclipse/share/keywords/900_OPM/B/BIOTCOEFF:2]: Done.

File: src/opm/output/eclipse/WriteInit.cpp [hunk 6]
```diff
@@ -444,6 +444,13 @@ namespace {
             {"PERMY" , ::Opm::UnitSystem::measure::permeability },
             {"PERMZ" , ::Opm::UnitSystem::measure::permeability },
             {"NTG"   , ::Opm::UnitSystem::measure::identity },
+            {"YMODULE"   , ::Opm::UnitSystem::measure::ymodule },
+            {"PRATIO"   , ::Opm::UnitSystem::measure::identity },
+            {"BIOTCOEF"   , ::Opm::UnitSystem::measure::identity },
+            {"POELCOEF"   , ::Opm::UnitSystem::measure::identity },//NB units not done
+            {"THELCOEF"   , ::Opm::UnitSystem::measure::identity },//NB units not done
+            {"THERMEXR"   , ::Opm::UnitSystem::measure::identity },
+            {"THCONR"   , ::Opm::UnitSystem::measure::identity } //NB units not done
```
[the remark this answers is not in the document]
Comment (hnil) [reply src/opm/output/eclipse/WriteInit.cpp:453]: Need to check again. The trouble is that introducing new units involve a lot of code changes do to the structure. Identity mean no conversion.

File: src/opm/input/eclipse/Schedule/BCProp.cpp (hunk 4 above)
Comment (hnil) [reply src/opm/input/eclipse/Schedule/BCProp.cpp:116]: I could have added stressxy also as value, but have not so fare. I thought it was better to make a proper stress i.e. 6 values xx,yy,zz,yz,xz,xy even if input did not support it.

File: src/opm/input/eclipse/Schedule/BCProp.cpp [hunk 7]
```diff
@@ -87,6 +103,38 @@ BCProp::BCFace::BCFace(const DeckRecord& record) :
...
+    if (const auto& P = record.getItem<BCKEY::FIXEDX>(); ! P.defaultApplied(0)) {
+        mechbcvaluetmp.fixeddir[0] = P.get<int>(0);
+    }
+    if (const auto& P = record.getItem<BCKEY::FIXEDY>(); ! P.defaultApplied(0)) {
+        mechbcvaluetmp.fixeddir[1] = P.get<int>(0);
+    }
+    if (const auto& P = record.getItem<BCKEY::FIXEDZ>(); ! P.defaultApplied(0)) {
+        mechbcvaluetmp.fixeddir[2] = P.get<int>(0);
+    }
+    mechbcvalue = mechbcvaluetmp;
```
[the remark this answers is not in the document]
Comment (hnil) [reply src/opm/input/eclipse/Schedule/BCProp.cpp:137]: I do not understand this. What can go wrong?

File: opm/input/eclipse/Schedule/BCProp.hpp [hunk 8]
```diff
@@ -52,18 +58,27 @@ enum class BCComponent {
...
+};
 
+    
 class BCProp {
 public:
 
     struct BCFace {
         int index;
         BCType bctype;
+        BCMECHType bcmechtype;
```
[the remark this answers is not in the document]
Comment (hnil) [reply opm/input/eclipse/Schedule/BCProp.hpp:74]: Done I think.

File: src/opm/input/eclipse/share/keywords/900_OPM/B/BC [hunk 9]
```diff
@@ -57,6 +57,62 @@
       "name": "TEMPERATURE",
       "value_type": "DOUBLE",
       "dimension": "Temperature"
-    }
+    },
```
[the remark this answers is not in the document]
Comment (hnil) [reply src/opm/input/eclipse/share/keywords/900_OPM/B/BC:60]: Probably separate pullrequest?

File: src/opm/input/eclipse/EclipseState/InitConfig/InitConfig.cpp (hunk 3 above)
Comment (hnil) [reply src/opm/input/eclipse/EclipseState/InitConfig/InitConfig.cpp:137]: Fixed.

File: src/opm/input/eclipse/EclipseState/SimulationConfig/BCConfig.cpp (hunk 1 above)
Comment (hnil) [reply src/opm/input/eclipse/EclipseState/SimulationConfig/BCConfig.cpp]: fixed.

File: src/opm/input/eclipse/EclipseState/InitConfig/InitConfig.cpp (hunk 2 above)
Comment (hnil) [reply src/opm/input/eclipse/EclipseState/InitConfig/InitConfig.cpp]: fixed

File: opm/input/eclipse/EclipseState/Grid/FieldProps.hpp [hunk 10]
```diff
@@ -137,6 +137,12 @@ static const std::unordered_map<std::string, keyword_info<double>> double_keywor
                                                                                       {"THCOIL",  keyword_info<double>{}.unit_string("Energy/AbsoluteTemperature*Length*Time")},
                                                                                       {"THCGAS",  keyword_info<double>{}.unit_string("Energy/AbsoluteTemperature*Length*Time")},
                                                                                       {"THCWATER",keyword_info<double>{}.unit_string("Energy/AbsoluteTemperature*Length*Time")},
+                                                                                      {"YMODULE", keyword_info<double>{}.unit_string("Giga*Pascal")},
```
[the remark this answers is not in the document]
Comment (hnil) [reply opm/input/eclipse/EclipseState/Grid/FieldProps.hpp:140]: ?

File: opm/input/eclipse/Schedule/BCProp.hpp [hunk 11]
```diff
@@ -76,6 +91,7 @@ class BCProp {
         {
             serializer(index);
             serializer(bctype);
+            //serializer(bcmechtype);
```
[the remark this answers is not in the document]
Comment (hnil) [reply opm/input/eclipse/Schedule/BCProp.hpp]: fixed?

File: opm/input/eclipse/Schedule/BCProp.hpp [hunk 12]
```diff
@@ -76,6 +91,7 @@ class BCProp {
         {
             serializer(index);
             serializer(bctype);
+            serializer(bcmechtype);
```
Comment (atgeirr) [inline opm/input/eclipse/Schedule/BCProp.hpp:94]: I think also mechbcvalue must be serialized. That is the only remaining problem I see in the PR.

File: src/opm/input/eclipse/Schedule/BCProp.cpp [hunk 13]
```diff
@@ -104,37 +104,48 @@ BCProp::BCFace::BCFace(const DeckRecord& record) :
...
+        all_ok = false;
     }
     if (const auto& P = record.getItem<BCKEY::FIXEDY>(); ! P.defaultApplied(0)) {
         mechbcvaluetmp.fixeddir[1] = P.get<int>(0);
     }
     if (const auto& P = record.getItem<BCKEY::FIXEDZ>(); ! P.defaultApplied(0)) {
         mechbcvaluetmp.fixeddir[2] = P.get<int>(0);
+        all_ok = false;
+    }
+    if(all_ok){
```
Comment (atgeirr) [inline src/opm/input/eclipse/Schedule/BCProp.cpp:146]: I believe the logic here is now 180 degree wrong... The all_ok variable should be named all_defaulted, then it should be
```
if (!all_defaulted) {
    mechbcvalue = mechbcvaluetmp;
}
```
(please also add space around the parenthesis above)

File: opm/input/eclipse/Schedule/BCProp.hpp (hunk 12 above)
Comment (atgeirr) [reply opm/input/eclipse/Schedule/BCProp.hpp:102]: This is still missing.

**Review by atgeirr — APPROVED** [review]
The changed logic we discussed is now a lot simpler and better, so I think this can be merged, as soon as the conflict in the unit system parts is resolved.

**Comment by hnil** [comment]
Tried to fix rebase. test need to be Rerunned.
@bska Is it a problem that addUnit(Ymodule... after addUnit(ContentDependent...

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by bska** [comment]
> Is it a problem that `addUnit(Ymodule...` after `addUnit(ContentDependent...`

That's not a problem.  The base units can be registered in any order.

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by hnil** [comment]
jenkins build this please

**Comment by akva2** [comment]
InitConfig::stressequil should be internalized, not stored as a DeckKeyword. in any case it needs to be serialized to fix the serialization tests.

**Comment by hnil** [comment]
Ok. Good that you find the serialization error.  What functionality do the internalization give?

**Comment by akva2** [comment]
Well for starters it makes it behave like any other things in the library. Deck items are used during parsing. These are internalized into the EclipseState and/or Schedule. This way the internals of the file format (item numbers etc) do not leak into the downstreams.

 In fact the Deck* classes were never intended to be part of the public API, it's sadly there still due to some ActionX things.

**Comment by akva2** [comment]
https://github.com/akva2/opm-common/tree/add_strequil implements internalization. i also changed to STREQUIL as suggested by @atgeirr to keep it at 8 chars. dunno if you want me to open the PR or you want to open it.

as a general rule several but smaller PRs are always far easier to process so I strongly suggest not to roll it into this one.

**Comment by atgeirr** [comment]
I assume this can be closed now?

**Pull request closed without merging**